pkg/sql/conn_executor.go

```diff
@@ -1626,28 +1626,42 @@ func (ex *connExecutor) execCopyIn(
 		ex.state.mon.Start(ctx, ex.sessionMon, mon.BoundAccount{} /* reserved */)
 		monToStop = ex.state.mon
 	}
-	cm, err := newCopyMachine(
-		ctx, cmd.Conn, cmd.Stmt, txnOpt, ex.server.cfg,
-
-		// resetPlanner
-		func(p *planner, txn *client.Txn, txnTS time.Time, stmtTS time.Time) {
-			// HACK: We're reaching inside ex.state and changing sqlTimestamp by hand.
-			// It is used by resetPlanner. Normally sqlTimestamp is updated by the
-			// state machine, but the copyMachine manages its own transactions without
-			// going through the state machine.
-			ex.state.sqlTimestamp = txnTS
-			ex.statsCollector = ex.newStatsCollector()
-			ex.statsCollector.reset(&ex.server.sqlStats, ex.appStats, &ex.phaseTimes)
-			ex.initPlanner(ctx, p)
-			ex.resetPlanner(ctx, p, txn, stmtTS, 0 /* numAnnotations */)
-		},
+	var cm copyMachineInterface
+	var err error
+	if cmd.Stmt.Table.TableName == fileUploadTable {
+		cm, err = newFileUploadMachine(cmd.Conn, cmd.Stmt, ex.server.cfg,
```

Comment:
Should we verify that cmd.Stmt.Columns is empty? It should probably be an error if it's not.

Reply:
We're breaking up the file content into rows of one column, so it shouldn't be empty as of right now. I can check that there is exactly one row though.

Reply:
I'm not sure we're talking about the same thing. I'm talking about the tree.CopyFrom node which could have columns. Like we should ensure a user doesn't do `COPY crdb_internal.file_upload(blah) TO ...` since the `blah` is completely ignored and may possibly do something the user doesn't expect.

Reply:
Ah I see, that's fair. I think currently, I actually do pass in a column name. That could cause some confusion since it's unused, will change it so that there is no column name, and check that it doesn't have it.